Continue button on unit overview should skip hidden lessons

Erin007 opened the pull request:
[LP-573](https://codedotorg.atlassian.net/browse/LP-573)

Currently the orange "Continue" button on student view of unit overview page doesn't work as expected when lessons are hidden.

ACTUAL: Students get a page telling them they're in the wrong place.
<img width="729" alt="Screen Shot 2019-09-09 at 4 17 59 PM" src="https://user-images.githubusercontent.com/12300669/64572764-8c210f80-d31d-11e9-8b30-938cf01fa988.png">

EXPECTED: Students go to the next visible lesson after the visible lesson they most recently completed

Hamms commented on dashboard/app/models/user.rb:
```diff
@@ -1233,6 +1273,16 @@ def script_level_hidden?(script_level)
 
+  def visible_script_levels(script)
+    visible_sls = []
+    script.script_levels.each do |sl|
```
note this could more simply be written as:

```ruby
def visible_script_levels(script)
  script.script_levels.select do |sl|
    !script_level_hidden?(sl)
  end
end

Hamms commented on dashboard/app/models/user.rb:
```diff
@@ -1138,6 +1138,46 @@ def has_activity?
+
+    # The user has not made any progress, return the first script_level
+    if ul_level_ids.empty?
+      return script.get_script_level_by_chapter(1)
```
shouldn't this return the first _visible_ script_level?

maddiedierker replied:
+1 to elijah's comment. you could have `sl_level_ids` be an array of visible script levels (rather than just IDs) to use on this line

Hamms commented on dashboard/app/models/user.rb:
```diff
@@ -1138,6 +1138,46 @@ def has_activity?
+    most_recent_ul = visible_user_levels.max_by(&:created_at)
+
+    # Find the script_level that goes with the most recent user_level
+    most_recent_sl = visible_script_levels(script).detect do |sl|
```
If we rewrite `visible_script_levels` to use a select, could this be `most_recent_sl = visible_script_levels(script).find_by(level_id: most_recent_ul.level_id)`?

Hamms commented on dashboard/app/models/user.rb:
```diff
@@ -1138,6 +1138,46 @@ def has_activity?
+      sl.chapter > most_recent_completed_chapter
+    end
+
+    next_unpassed_visible_progression_sl = later_visible_sls.min_by(&:chapter)
```
love the use of min_ and max_by!

Hamms commented on dashboard/app/controllers/script_levels_controller.rb:
```diff
@@ -282,7 +282,7 @@ def next_script_level
 
   def user_or_session_level
     if current_user
-      current_user.next_unpassed_progression_level(@script).try(:or_next_progression_level)
```
should the old method be deprecated?

Erin007 replied:
I wondered about that too, and opened [LP-733](https://codedotorg.atlassian.net/browse/LP-733) to track this clean up. 

We use `next_unpassed_progression_level` in a few other places (home controller, maker controller and a user helper). It's likely that it's causing similar bugs with hidden lessons in those places but I want to investigate a little deeper so I don't break something along the way.  After I check those uses, I'll decide if `next_unpassed_progression` level should be replaced with `next_unpassed_visible_progression_level`, combined with an exclude_hidden param or if we still need it separate for valid use cases.

Hamms commented on dashboard/app/controllers/script_levels_controller.rb:
```diff
@@ -282,7 +282,7 @@ def next_script_level
   def user_or_session_level
     if current_user
-      current_user.next_unpassed_progression_level(@script).try(:or_next_progression_level)
+      current_user.next_unpassed_visible_progression_level(@script).try(:or_next_progression_level)
```
is the `or_next_progression_level` bit the part that handles the "we're on the last level in the script, so there _is_ no next progression level" logic?

Erin007 replied:
Yes, but it was plagued by a similar problem. If you complete the last level without completing all of the previous ones, and click continue on the unit overview page, you'll end up on the first level of the script _even if it's hidden_, in which case you get the "your teacher wasn't expecting you here" message. In the latest change, I removed `or_next_progression_level` and account for the last level in a progression in `next_unpassed_visible_progression_level`.

Hamms commented on dashboard/app/models/user.rb:
```diff
@@ -1138,6 +1138,46 @@ def has_activity?
+
+    next_unpassed_visible_progression_sl = later_visible_sls.min_by(&:chapter)
+
+    next_unpassed_visible_progression_sl
```
does there need to be fallback logic here, akin to https://github.com/code-dot-org/code-dot-org/blob/9be5d1e0c0752205526ea796045c7b7d358919e2/dashboard/app/models/user.rb#L1162-L1164 ?

Erin007 replied:
With the latest changes I handle this logic within the `next_unpassed_visible_progression_level` method, which is even more of a beast so I don't know if that was the best decision...

maddiedierker commented on dashboard/app/models/user.rb:
```diff
@@ -1138,6 +1138,46 @@ def has_activity?
   end
 
+  # Returns the next visible script_level for the next progression level in the given
+  # script that hasn't yet been passed, starting its search at the last level we submitted
```
super nit:
```suggestion
  # script that hasn't yet been passed, starting its search at the last level the user submitted
```

Erin007 commented:
> One meta-comment; it certainly makes sense for this method to live here given our current approach to organizing code related to users (ie, dump everything into user.rb) but we're [working toward a better organizational system for this model](https://github.com/code-dot-org/code-dot-org/pull/30444).
> 
> Ideally both this code and the existing progression level logic would live in a Policy Object; that extraction may or may not be out of scope for this PR, but we should at the very least add a work item to track performing it.

Agreed! tracking in [LP-734](https://codedotorg.atlassian.net/browse/LP-734)

Erin007 commented:
@madelynkasula @Hamms I incorporated your feedback, handled some edge cases not previously addressed and added more test coverage. Can you take another look please?

maddiedierker commented on dashboard/app/models/user.rb:
```diff
@@ -1138,6 +1138,63 @@ def has_activity?
+
+    first_visible_level = visible_sls.min_by(&:chapter)
+
+    completed_all_visible_levels = visible_incomplete_level_ids.empty? && !completed?(script)
```
should we check `completed?(script)` before we do anything else? it doesn't look like we're handling that case in this method. 

if we should, i think we should calculate "completed" once at the beginning of the method as well, since we currently have to call this helper a few times, where it will hit the database each time. something like:

```
is_complete = completed?(script)
return [congrats page? idk] if is_complete
```

Erin007 replied:
the `next` method in `script_level_controller` handles the redirect to congrats if the script is completed. So we should never get to `build-script_level_path(next_script_level)` which calls `user_or_session_level` 
 which calls `current_user_next_unpassed_progression_level`

Erin007 replied:
https://github.com/code-dot-org/code-dot-org/blob/d5c83fca49f75958603eacf56f184465b98d053c/dashboard/app/controllers/script_levels_controller.rb#L72

Erin007 replied:
but there's no point in doing all of the work for this method if the script is complete... so I threw it all in an `unless script.complete?`

maddiedierker replied:
ahh, okay cool. since that conditional is around the entire method, i think it's cleaner to add an immediate return statement at the top to reduce nesting:

```
return nil if script.complete?
[...rest of method]
```

maddiedierker commented on dashboard/app/models/user.rb:
```diff
@@ -1138,6 +1138,63 @@ def has_activity?
+
+    # The user has not made any progress or has completed all visible levels
+    # but not the entire script, return the first visible script_level
+    if ul_level_ids.empty? || completed_all_visible_levels
```
let's move calculating `visible_incomplete_sls` and `first_visible_incomplete_level` below this conditional, so that we're only calculating those values if we really need to

Hamms commented on dashboard/app/models/user.rb:
```diff
@@ -1138,6 +1138,67 @@ def has_activity?
+      # but not the entire script, return the first visible script_level
+      if ul_level_ids.empty? || completed_all_visible_levels
+        return first_visible_level
+      end
```
nit: this (and the `most_recent_sl == last_visible_level` check below) can just be `return first_visible_level if ul_level_ids.empty? || completed_all_visible_levels`

Hamms commented on dashboard/app/models/user.rb:
```diff
@@ -1138,6 +1138,67 @@ def has_activity?
+      # The user has completed the last level in the progress, but not all
+      # previous levels, return the first visible incomplete script_level
+      if most_recent_sl == last_visible_level
+        return first_visible_incomplete_level ? first_visible_incomplete_level : first_visible_level
```
this could also be

```suggestion
        return first_visible_incomplete_level || first_visible_level
```

Erin007 commented:
@madelynkasula @Hamms The monstrosity has passing tests and I've taken your suggestions (thank you!).  Anything else look problematic?

maddiedierker commented:
no, i think this is good to merge! thanks for your patience and persistence on this, it was a doozy

Erin007 commented:
> > One meta-comment; it certainly makes sense for this method to live here given our current approach to organizing code related to users (ie, dump everything into user.rb) but we're [working toward a better organizational system for this model](https://github.com/code-dot-org/code-dot-org/pull/30444).
> > Ideally both this code and the existing progression level logic would live in a Policy Object; that extraction may or may not be out of scope for this PR, but we should at the very least add a work item to track performing it.
> 
> Agreed! tracking in [LP-734](https://codedotorg.atlassian.net/browse/LP-734)

Getting ready to tackle this - thinking about a set up following #30752 that ends up with being able to call: `Queries::UserProgress.next_unpassed_visible_progression_level(user, script)` @Hamms , am I on the right track?  I'm thinking Query::UseProgress would be  where a handful of the relevant methods would live (e.g. `completeled_progression_levels`, `user_progress_by_stage`)...

Hamms commented:
that sounds absolutely ideal!

As mentioned in that PR, I had also been playing around with some ideas for how we could build these Query Objects in such as way as to be easily chainable or composable, but didn't ever land on something that worked really well. As you start digging into this feature, if you start to suspect that it might benefit from some of that functionality please lemme know! I'd love to take another crack at it
